## Assignment: 2

Topic: List and Tuples

#### <span style='color:purple '>1. What is a List?<span>

List is sequential data structure is python that is mutable, changeble, iterable,and ordered.

#### <span style='color:purple '>2. What is a Tuple?<span>

Tuple is collection of python objects seperated by commas, In someway tuple in similar to list like indexing and repeatation but unlike list tuple is immutable.

#### <span style='color:purple '>3. What is the difference between List and Tuple? <span>

    List                       Tuple
-List is mutable            -Tuple is immutable
-List has variable length   -Tuple has fixed length
-List is more functional    -Tuple is less functional
-Requires more memory       -Requires less memory
-slower than tuple          -faster that list

#### <span style='color:purple '>4. Python Program to find the largest element in the list <span>

In [2]:
#finding the element with max value from list

l1=[23,56,78,54,21,45]
print(max(l1))

78


In [20]:
def large_element(list):
    d={}
    for i in list:
        d.update({i:len(i)})
    d2=sorted(d.items(),key=lambda kv:kv[1])
    return d2[-1][0]

In [21]:
list_1=['Sarika','Megha','Ritu', 'Aishwarya','Ketaki','Riya']

large_element(list_1)

'Aishwarya'

#### <span style='color:purple '>5. Python program to interchange first and last elements in a list.<span>

In [22]:
list_1=['Sarika','Megha','Ritu', 'Aishwarya','Ketaki','Riya']

first_ele=list_1[0]
last_ele=list_1[-1]

list_1[0]=last_ele
list_1[-1]=first_ele
print(list_1)

['Riya', 'Megha', 'Ritu', 'Aishwarya', 'Ketaki', 'Sarika']


In [30]:
l1=[1,2,3,4,5,6,7,8,9]
fe=l1[0:1]
le=l1[-1:]

new_list=le+l1[1:-1]+fe
print(new_list)

[9, 2, 3, 4, 5, 6, 7, 8, 1]


#### <span style='color:purple '>6. Python program to swap two elements in a list <span>

In [34]:
def swap_elements(list,ele1,ele2):
    index_1=list.index(ele1)
    index_2=list.index(ele2)
    
    list[index_1]=ele2
    list[index_2]=ele1
    
    return list

In [35]:
list_1=['Sarika','Megha','Ritu', 'Aishwarya','Ketaki','Riya']
swap_elements(list_1,'Megha','Riya')

['Sarika', 'Riya', 'Ritu', 'Aishwarya', 'Ketaki', 'Megha']

#### <span style='color:purple '>7. Python program to Reverse a List <span>

In [1]:
l=[1,2,3,4,5]
l[::-1]

[5, 4, 3, 2, 1]

#### <span style='color:purple '>8. Python program to count occurrences of an element in a list <span>

In [37]:
def count_ele(list,element):
    count_=list.count(element)
    return "Count of element is:", count_

list1=[22,33,44,55,22,12]

count_ele(list1,22)

('Count of element is:', 2)

#### <span style='color:purple '>9. Python program to find the sum of elements in a list <span>

In [38]:
list1=[22,33,44,55,22,12]

print('Sum of elements in list is:',sum(list1))

Sum of elements in list is: 188


#### <span style='color:purple '>10. Python program to Multiply all numbers in the list <span>

In [43]:
list1=[22,33,44,55,22,12]

multi=1
for i in list1:
    multi *= i

print('Multiplication of all numbers in list is:',multi)

Multiplication of all numbers in list is: 463826880


#### <span style='color:purple '>11. What are the ways to find the length of a list <span>

In [44]:
#using len function
list1=[22,33,44,55,22,12]
len(list1)

6

In [45]:
list1=[22,33,44,55,22,12]
length=0

for i in list1:
    length+=1
    
print('Length:',length)

Length: 6


#### <span style='color:purple '>12. Python program to find the smallest and largest number in a list (Without min-max function) <span>

In [48]:
def min_max(list):
    list2=sorted(list)
    print('Min element of list:',list2[0])
    print('Max element of list:',list2[-1])
    return

list1=[32,56,43,12,76,52]
min_max(list1)

Min element of list: 12
Max element of list: 76


#### <span style='color:purple '>13. Python Program to find the area of a circle <span>

In [52]:
from math import pi

radius=int(input())
print('Area of circle:', round(pi*radius**2,2))

4
Area of circle: 50.27


#### <span style='color:purple '>14. Take inputs from the user to make a list. Again take one input from the user and search it in the list and delete that element, if found. Iterate over a list using for loop.<span>

In [56]:
l1=[]
length=int(input('Enter the length of list:'))

for i in range(1,length+1):
    element=int(input(f'Enter {i} th element:'))
    l1.append(element)

print(l1)

el2=int(input('Enter element to be deleted:'))

if el2 in l1:
    l1.remove(el2)
    print('updated list',l1)
else:
    print('element is not available in list')



Enter the length of list:3
Enter 1 th element:7
Enter 2 th element:8
Enter 3 th element:9
[7, 8, 9]
Enter element to be deleted:2
element is not available in list


In [57]:
l1=[]
length=int(input('Enter the length of list:'))

for i in range(1,length+1):
    element=int(input(f'Enter {i} th element:'))
    l1.append(element)

print(l1)

el2=int(input('Enter element to be deleted:'))

if el2 in l1:
    l1.remove(el2)
    print('updated list',l1)
else:
    print('element is not available in list')

Enter the length of list:4
Enter 1 th element:11
Enter 2 th element:22
Enter 3 th element:33
Enter 4 th element:44
[11, 22, 33, 44]
Enter element to be deleted:22
updated list [11, 33, 44]


#### <span style='color:purple '>15. You are given a list of integer elements. Make a new list that will store a square of elements of the previous list. (With and without list comprehension) <span>
i. Input_list = [2,5,6,12]
ii. Output_list = [4,25,36,144]


In [59]:
#without list comprehension
Input_list = [2,5,6,12]
Output_list=[]

for i in Input_list:
    Output_list.append(i**2)
    
print(Output_list)

[4, 25, 36, 144]


In [60]:
#with list comprehension
Input_list = [2,5,6,12]
Output_list=[i**2 for i in Input_list]

print(Output_list)

[4, 25, 36, 144]


#### <span style='color:purple '>16. WAP to create two lists, one containing all even numbers and the other containing all odd numbers between 0 to 151 <span>

In [66]:
even_num=[i for i in range (152) if i%2==0]
odd_num=[i for i in range (152) if i%2!=0]

print('Even numbers:\n',even_num)
print('-'*100)
print('Odd numbers:\n',odd_num)

Even numbers:
 [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150]
----------------------------------------------------------------------------------------------------
Odd numbers:
 [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151]


In [67]:
def seperate_odd_even(num):
    even_num=[]
    odd_num=[]
    
    for i in range(num+1):
        if i%2==0:
            even_num.append(i)
        else:
            odd_num.append(i)
            
    print('Even numbers:\n',even_num)
    print('Odd numbersL\n',odd_num)
    return

seperate_odd_even(151)


Even numbers:
 [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150]
Odd numbersL
 [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151]


#### <span style='color:purple '>17. Python program to Count Even and Odd numbers in a List<span>

In [68]:
lis1=[32,45,67,98,65,43,13,5,6,76]

even_count,odd_count=0,0

for i in lis1:
    if i%2==0:
        even_count+=1
    else:
        odd_count+=1
        
print('Even numbers in list:',even_count)
print('Odd numbers in list:',odd_count)

Even numbers in list: 4
Odd numbers in list: 6


#### <span style='color:purple '>18. WAP to make new lists, containing only numbers which are divisible by 4, 6, 8, 10, 3, 5, 7, and 9 in separate lists for range(0,151) <span>

In [71]:
divisible_by_3=[i for i in range(1,152) if i%3==0]
divisible_by_4=[i for i in range(1,152) if i%4==0]
divisible_by_5=[i for i in range(1,152) if i%5==0]
divisible_by_6=[i for i in range(1,152) if i%6==0]
divisible_by_7=[i for i in range(1,152) if i%7==0]
divisible_by_8=[i for i in range(1,152) if i%8==0]
divisible_by_9=[i for i in range(1,152) if i%9==0]
divisible_by_10=[i for i in range(1,152) if i%10==0]

print('Numbers divisibleed by 3:\n',divisible_by_3)
print('-'*100)
print('Numbers divisibleed by 4:\n',divisible_by_4)
print('-'*100)
print('Numbers divisibleed by 5:\n',divisible_by_5)
print('-'*100)
print('Numbers divisibleed by 6:\n',divisible_by_6)
print('-'*100)
print('Numbers divisibleed by 7:\n',divisible_by_7)
print('-'*100)
print('Numbers divisibleed by 8:\n',divisible_by_8)
print('-'*100)
print('Numbers divisibleed by 9:\n',divisible_by_9)
print('-'*100)
print('Numbers divisibleed by 10:\n',divisible_by_10)
print('-'*100)

Numbers divisibleed by 3:
 [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150]
----------------------------------------------------------------------------------------------------
Numbers divisibleed by 4:
 [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 144, 148]
----------------------------------------------------------------------------------------------------
Numbers divisibleed by 5:
 [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100, 105, 110, 115, 120, 125, 130, 135, 140, 145, 150]
----------------------------------------------------------------------------------------------------
Numbers divisibleed by 6:
 [6, 12, 18, 24, 30, 36, 42, 48, 54, 60, 66, 72, 78, 84, 90, 96, 102, 108, 114, 120,

#### <span style='color:purple '>19. From a list containing ints, strings, and floats, make three lists to store them separately.<span>

In [74]:
list1=[34,6,7,'science',8.76,32.1,5,66.7,8,'Python']

strings,floats,ints=[],[],[]

for i in list1:
    if type(i)==int:
        ints.append(i)
    elif type(i)==float:
        floats.append(i)
    else:
        strings.append(i)

print(ints)
print(floats)
print(strings)


[34, 6, 7, 5, 8]
[8.76, 32.1, 66.7]
['science', 'Python']


#### <span style='color:purple '>20. What’s The Difference Between The Python append() and extend() Methods? <span>

append : use to add single value or list to the list at last index
extend: use to add only list to the list at last index

#### <span style='color:purple '>21. Write a Python program to append a list to the second list <span>

In [76]:
l1=[12,13,14]
l2=[21,32,43]

l1.append(l2)
print(l1)

[12, 13, 14, [21, 32, 43]]


#### <span style='color:purple '> 22. Write a Python program to find the third-largest number in a list<span>

In [80]:
l1=[34,55,67,4,32,4,56,7,6,53]
l2=sorted(l1)
print('Third largest element:',l2[-3])

Third largest element: 55


#### <span style='color:purple '>23. Write a Python program to get the frequency of the elements in a list.<span>

In [90]:
list1= ['f','d','v','n','g','f','j','s','y','d','s','a','t','d','s','w','f','g','a','r','v','d','y']
d={}

for i in list1:
    d.update({i:list1.count(i)})

print(d)

{'f': 3, 'd': 4, 'v': 2, 'n': 1, 'g': 2, 'j': 1, 's': 3, 'y': 2, 'a': 2, 't': 1, 'w': 1, 'r': 1}


In [94]:
list1= ['f','d','v','n','g','f','j','s','y','d','s','a','t','d','s','w','f','g','a','r','v','d','y']

element=input('Enter your element:')

if element in list1:
    print(f'count of {element} in list is {list1.count(element)}')
else:
    print('your element is not in list')

Enter your element:s
count of s in list is 3


In [95]:
list1= ['f','d','v','n','g','f','j','s','y','d','s','a','t','d','s','w','f','g','a','r','v','d','y']

element=input('Enter your element:')

if element in list1:
    print(f'count of {element} in list is {list1.count(element)}')
else:
    print('your element is not in list')

Enter your element:k
your element is not in list


#### <span style='color:purple '>24. Write a Python program to check whether a list contains a sublist <span>

In [101]:
def check_sublist(list,sublist):
    result=False
    for i in range(len(list)):
        if list[i:i+len(sublist)]==sublist:
            result=True
            break
    print(f'{sublist} is part of {list}?', result)
        
list1=[45,67,32,89,90,64,32,12,56]
sublist=[89,90,64]

check_sublist(list1,sublist)

[89, 90, 64] is part of [45, 67, 32, 89, 90, 64, 32, 12, 56]? True


In [102]:
list1=[45,67,32,89,90,64,32,12,56]
sublist=[89,32,64]

check_sublist(list1,sublist)

[89, 32, 64] is part of [45, 67, 32, 89, 90, 64, 32, 12, 56]? False


#### <span style='color:purple '>25. Write a Python program to generate all sublists of a list <span>

In [115]:
def creat_sublist(l):
    sub_lists=[[]]
    for i in range(len(l)+1):
        for j in range(i):
            sub_lists.append(l[j: i])
    return sub_lists

lis=[1,2,3]
creat_sublist(lis)

[[], [1], [1, 2], [2], [1, 2, 3], [2, 3], [3]]

#### <span style='color:purple '>26. Write a Python program to find common items from two lists <span>

In [116]:
l1=[43,65,78,54,23,12,45,76,87]
l2=[56,23,55,78,65,3,2,12,32,56,87]

common=[]
for i in l1:
    if i in l2:
        common.append(i)

print(common)

[65, 78, 23, 12, 87]


#### <span style='color:purple '>27. How to flatten a list in python? <span>

In [124]:
#falttening a list using list comprehension

l1=[[1,2],[3,4,5],[6,7],[8,9]]

flat_list=[i for ls in l1 for i in ls]
print(flat_list)


[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [125]:
#using itertools

import itertools

l1=[[1,2],[3,4,5],[6,7],[8,9]]
flat_list=list(itertools.chain(*l1))

print(flat_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [128]:
#using numpy library
import numpy as np

l1=[[1,2],[3,4,5],[6,7],[8,9]]
flat_list=list(np.concatenate(l1).flat)
print(flat_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [129]:
#suing sum function

l1=[[1,2],[3,4,5],[6,7],[8,9]]
f_l=sum(l1,[])
print(f_l)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


#### <span style='color:purple '>28. How to sort a list in ascending and descending order without using the sort function? <span>

In [134]:
list1=[45,67,32,89,90,64,32,12,56]

for i in range(len(list1)):
    for j in range(i+1,len(list1)):
        if list1[i]>list1[j]:
            list1[i],list1[j]=list1[j],list1[i]

print('List in ascending order:',list1)

List in ascending order: [12, 32, 32, 45, 56, 64, 67, 89, 90]


In [135]:
list1=[45,67,32,89,90,64,32,12,56]

for i in range(len(list1)):
    for j in range(i+1,len(list1)):
        if list1[i]<list1[j]:
            list1[i],list1[j]=list1[j],list1[i]

print('List in descending order:',list1)

List in descending order: [90, 89, 67, 64, 56, 45, 32, 32, 12]


#### <span style='color:purple '>29. How to sort a tuple?<span>

In [140]:
tup=(34,56,78,21)

print('Tuple sorted in ascending order:',tuple(sorted(tup)))
print('Tuple sorted in descending order:',tuple(sorted(tup,reverse=True)))

Tuple sorted in ascending order: (21, 34, 56, 78)
Tuple sorted in descending order: (78, 56, 34, 21)


#### <span style='color:purple '>30. Write a Python program to convert a list of multiple integers into a single
integer <span>
a. [11, 33, 50] >>> 113350

In [141]:
l=[11,33,50]
st=''

for i in l:
    s=str(i)
    st+=s
print(st)

113350


#### <span style='color:purple '>31. Difference between del and clear? <span>

del : It is a keyword but not a function. It delets the list selected to be deleted with all its elements from the memory location. Thus, printing it again would show an error of undefined variable.

clear : It is a function of the list. It will clear all the elements from the list, thus giving an empty list as an output. It will be stored in the memory but will be empty.

#### <span style='color:purple '>32. Difference between remove and pop?<span>

remove : It is used to remove the first element from the list by the use of "item".

In [144]:
l=[5,43,2,3,5,6]
l.remove(5) #item
l

[43, 2, 3, 5, 6]

pop : It is used to remove the element from the list by using the index of the element in the list. If index is not defined, it will automatically pop the last value in the list.

In [145]:
l=[5,43,2,3,5,6]
l.pop(5) #index
l

[5, 43, 2, 3, 5]

#### <span style='color:purple '>33. Difference between indexing and Slicing? <span>

indexing : Indexing means referring to an element of an iterable by its position within the iterable.

slicing : Slicing means getting a subset of elements from an iterable based on their indices.

#### <span style='color:purple '>34. Difference between sort and sorted?<span>

sort() : sort() function is very similar to sorted() but unlike sorted it returns nothing and makes changes to the original sequence. Moreover, sort() is a method of list class and can only be used with lists.

sorted() : sorted() method sorts the given sequence either in ascending order or in descending order and always return the a sorted list. This method does not effect the original sequence.

#### <span style='color:purple '> 35. Difference between reverse and reversed? <span>

reverse() : It does an in-place reversal and does not return a value.

reversed() : It does not change the list, but returns a reverse iterable object.

#### <span style='color:purple '> 36. Difference between copy and deep copy? <span>

copy : Also known as shallow copy. This function copies the elements of the lists as references of the original list.

deep copy : This function is related to nested lists. It copies all the elements of the lists including the lists of lists, thus a recursive copy.

#### <span style='color:purple '>37. How to check whether the list is empty or not?<span>

By using the len(list1) function : If the length of the list is 0, it is empty.

#### <span style='color:purple '>38. How to concatenate two lists? <span>

In [146]:
# By using the arithmetic operator such as "+".

l1=[1,2,3]
l2=[4,5,6]
print(l1+l2)

[1, 2, 3, 4, 5, 6]


In [147]:
# By using the append() function.
l1=[1,2,3]
l2=[4,5,6]
l1.append(l2) #appends given list as last element
print(l1)

[1, 2, 3, [4, 5, 6]]


In [148]:
# By using the extend() function.
l1=[1,2,3]
l2=[4,5,6]
l1.extend(l2)
print(l1)

[1, 2, 3, 4, 5, 6]


#### <span style='color:purple '> 39. How to find the occurrences of an element in the python list? <span>

By using the Count function.

By using the for loop and the len() function together.

#### <span style='color:purple '>40. How to flatten a list in python?<span>

Using a list comprehension : A Python list comprehension defines a list based on the contents of an existing list. You can customize each element that goes into your new list by specifying conditions or changes within the comprehension.

Using a nested for loop : Nested for loop is another way of saying a for loop within a for loop

In [ ]:
#end